test/Docker.DotNet.Tests/IContainerOperationsTests.cs

```diff
@@ -0,0 +1,558 @@
+        [InlineData(5)]
+        [InlineData(10)]
+        public async Task CreateContainerAsync_TimeoutExpires_Fails(int millisecondsTimeout)
+        {
+            using var dockerClientWithTimeout = _dockerClientConfiguration.CreateClient();
+
+            dockerClientWithTimeout.DefaultTimeout = TimeSpan.FromMilliseconds(millisecondsTimeout);
+
+            _output.WriteLine($"Time available for CreateContainer operation: {millisecondsTimeout} ms'");
+
+            var timer = new Stopwatch();
+            timer.Start();
+
+            var createContainerTask = dockerClientWithTimeout.Containers.CreateContainerAsync(
+                new CreateContainerParameters
+                {
+                    Image = _imageId,
+                    Name = Guid.NewGuid().ToString(),
+                },
+                _cts.Token);
+
+            _ = await Assert.ThrowsAsync<OperationCanceledException>(() => createContainerTask);
+
+            timer.Stop();
+            _output.WriteLine($"CreateContainerOperation finished after {timer.ElapsedMilliseconds} ms");
+
+            Assert.True(createContainerTask.IsCanceled);
+            Assert.True(createContainerTask.IsCompleted);
+        }
+
+        [Fact]
+        public async Task GetContainerLogs_Tty_False_ReadsLogs()
+        {
+            using var containerLogsCts = new CancellationTokenSource(TimeSpan.FromSeconds(50));
+            var logList = new List<string>();
+
+            var createContainerResponse = await _dockerClient.Containers.CreateContainerAsync(
+                new CreateContainerParameters()
+                {
+                    Image = _imageId,
+                    Name = Guid.NewGuid().ToString(),
+                    Tty = false
+                },
+                _cts.Token
+            );
+
+            await _dockerClient.Containers.StartContainerAsync(
+                createContainerResponse.ID,
+                new ContainerStartParameters(),
+                _cts.Token
+            );
+
+            await _dockerClient.Containers.GetContainerLogsAsync(
+                createContainerResponse.ID,
+                new ContainerLogsParameters
+                {
+                    ShowStderr = true,
+                    ShowStdout = true,
+                    Timestamps = true,
+                    Follow = false
+                },
+                containerLogsCts.Token,
+                new Progress<string>((m) => { logList.Add(m); _output.WriteLine(m); })
+            );
+
+            await _dockerClient.Containers.StopContainerAsync(
+                createContainerResponse.ID,
+                new ContainerStopParameters(),
+                _cts.Token
+                );
+
+            _output.WriteLine($"Line count: {logList.Count}");
+
+            Assert.NotEmpty(logList);
+        }
+
+        //  @dgvives: This test doesn't work on xUnit test project. Execution never finishes.
+        //            Same code on consoleApp works just fine
```

Comment:
This may suggest a deadlock bug in the code, rather than an xUnit problem. It's common for async code to work in Console apps but break elsewhere, due to the difference in synchronization contexts. Unfortunately, I'm not an expert in how sync contexts and schedulers and such work, so I don't know where to suggest looking.

Reply:
I see, I'm creating some sort of web app to test this. Will update with test results

Reply:
My bad, failing test wasn't **IContainerOperations.GetContainerLogs_Tty_True_StreamReadsLogs** but **IContainerOpreations.GetContainerStatsAsync_StreamStats**

I didn't notice it because it works fine on build agent.

Reply:
JsonReader.ReadAsync can't be cancelled once serialization has started.
Looking for an alternative now